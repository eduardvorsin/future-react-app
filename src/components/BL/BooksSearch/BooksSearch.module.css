.books-search {
	/* .books-search styles */
}

.books-search__search-bar {
	margin-bottom: 20px;
	max-width: 100%;
}

.books-search__select {
	margin-bottom: 15px;
}

.books-search__category-select {
	/* .books-search__category-select styles */
}

.books-search__sort-select {
	/* .books-search__sort-select styles */
}