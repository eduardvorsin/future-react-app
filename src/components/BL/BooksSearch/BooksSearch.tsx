import React, { FC, useCallback, useContext, useMemo, useState } from 'react';
import { useNavigate } from 'react-router-dom';
import { useTranslation } from 'react-i18next';
import fetchBooks from '../../../store/thunks/fetchBooks/fetchBooks';
import useAppDispatch from '../../../hooks/useAppDispatch/useAppDispatch';
import { setSearchBookName, setSearchCategory, setSearchSortOrder, setSearchBy } from '../../../store/bookSlice/bookSlice';
import SearchBar from '../../UI/SearchBar/SearchBar';
import classes from './BooksSearch.module.css';
import useAppSelector from '../../../hooks/useAppSelector/useAppSelector';
import { createOptionsFromValues, resetVerticalScrollPosition } from '../../../helpers/helpers';
import { BookCategories, BookSearchBy, BookSortOrder, CategoryOption, SearchByOption, SorterOrderOption } from '../../../API/bookTypes';
import { LanguageContext } from '../../../contexts/LanguageContext';
import { MemoSelect } from '../../UI/Select/Select';

const categories: BookCategories[] = ['all', 'art', 'biography', 'computers', 'history', 'medical', 'poetry'];
const sortingOrder: BookSortOrder[] = ['relevance', 'newest'];
const searchBy: BookSearchBy[] = ['intitle', 'inauthor', 'inpublisher', 'isbn'];

type BooksSearchProps = {
  className?: string,
}

const BooksSearch: FC<BooksSearchProps> = ({
  className,
}) => {
  const { value: language } = useContext(LanguageContext);
  const { t } = useTranslation();
  const dispatch = useAppDispatch();
  const navigate = useNavigate();
  const { searchOptions } = useAppSelector((state) => state.books);

  const categoriesOptions = useMemo(() => createOptionsFromValues(categories, 'bookSearch.categories') as CategoryOption[], [language]);
  const sortOrderOptions = useMemo(() => createOptionsFromValues(sortingOrder, 'bookSearch.sortOrder') as SorterOrderOption[], [language]);
  const searchByOptions = useMemo(() => createOptionsFromValues(searchBy, 'bookSearch.searchBy') as SearchByOption[], [language]);

  const [searchValue, setSearchValue] = useState<string>('');
  const [category, setCategory] = useState<BookCategories>('all');
  const [sortOrder, setSortOrder] = useState<BookSortOrder>('relevance');
  const [findBy, setFindBy] = useState<BookSearchBy>('intitle');

  const changeHandler: React.ChangeEventHandler<HTMLInputElement> = (e) => {
    setSearchValue(e.target.value);
  };

  const searchHandler: () => void = () => {
    dispatch(setSearchBookName(searchValue));
    dispatch(fetchBooks({
      page: 0,
      bookName: searchValue,
      sortOrder,
      category,
      searchBy: findBy,
    }));

    resetVerticalScrollPosition();
    navigate('/books');
  };

  const clearHandler: () => void = () => {
    setSearchValue('');
  };

  const sortOrderChangeHandler: React.ChangeEventHandler<HTMLSelectElement> = useCallback((e) => {
    const value = e.currentTarget.value as BookSortOrder;

    dispatch(setSearchSortOrder(value));
    setSortOrder(value);
  }, [sortOrder]);

  const categoryChangeHandler: React.ChangeEventHandler<HTMLSelectElement> = useCallback((e) => {
    const value = e.currentTarget.value as BookCategories;

    dispatch(setSearchCategory(value));
    setCategory(value);
  }, [category]);

  const searchByChangeHandler: React.ChangeEventHandler<HTMLSelectElement> = useCallback((e) => {
    const value = e.currentTarget.value as BookSearchBy;

    dispatch(setSearchBy(value));
    setFindBy(value);
  }, [setFindBy]);

  const booksSearchClasses = [
    classes['books-search'],
    className,
  ].join(' ');

  return (
    <div
      className={booksSearchClasses}
    >
      <SearchBar
        className={classes['books-search__search-bar']}
        value={searchValue}
        onClear={clearHandler}
        onSearch={searchHandler}
        onChange={changeHandler}
      />

      <MemoSelect
        className={`${classes['books-search__category-select']} ${classes['books-search__select']}`}
        value={category}
        labelText={t('bookSearch.categoriesLabel', { ns: 'homePage' })}
        id='categories'
        onChange={categoryChangeHandler}
        options={categoriesOptions}
      />

      <MemoSelect
        className={`${classes['books-search__sort-select']} ${classes['books-search__select']}`}
        value={sortOrder}
        labelText={t('bookSearch.sortOrderLabel', { ns: 'homePage' })}
        id='sortingBy'
        onChange={sortOrderChangeHandler}
        options={sortOrderOptions}
      />

      <MemoSelect
        className={`${classes['books-search__find-by-select']} ${classes['books-search__select']}`}
        value={findBy}
        labelText={t('bookSearch.searchByLabel', { ns: 'homePage' })}
        id='findBy'
        onChange={searchByChangeHandler}
        options={searchByOptions}
      />
    </div>
  );
};

export default BooksSearch;
