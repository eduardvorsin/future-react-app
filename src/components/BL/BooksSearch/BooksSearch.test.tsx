import React from 'react';
import userEvent from '@testing-library/user-event';
import { render, screen, waitFor } from '@testing-library/react';
import { RouterProvider, createMemoryRouter } from 'react-router-dom';
import BooksSearch from './BooksSearch';
import i18n from '../../../localization/i18next';

const mockFn = jest.fn();
const router = createMemoryRouter([{
  path: '/',
  element: (
    <BooksSearch
      testId='test-book-search'
      onSearch={mockFn}
    />
  ),
}]);

describe('BooksSearch tests', () => {
  afterEach(async () => {
    await waitFor(() => router.navigate('/'));
  });

  it('is rendered correctly', () => {
    render(
      <RouterProvider router={router} />,
    );

    expect(screen.getByTestId<HTMLDivElement>('test-book-search')).toBeInTheDocument();
  });

  it('when you enter a value in the search field and click on the find button, the mock function should be called', async () => {
    const user = userEvent.setup();
    render(
      <RouterProvider router={router} />,
    );

    await waitFor(() => {
      user.type(screen.getByRole<HTMLInputElement>('textbox'), '123');
    });

    await waitFor(() => {
      user.click(screen.getByRole<HTMLButtonElement>('button', { name: /find/i }));
    });

    await waitFor(() => {
      expect(mockFn).toHaveBeenCalledTimes(1);
    });
  });

  it('is a basic snapshot', () => {
    render(
      <RouterProvider router={router} />,
    );

    expect(screen.getByTestId<HTMLDivElement>('test-book-search')).toMatchSnapshot();
  });
});

describe('BooksSearch integration tests', () => {
  afterEach(async () => {
    await waitFor(() => router.navigate('/'));
  });

  it('the component with the text in English is displayed correctly', () => {
    const localizedCategoriesLabel = i18n.t('bookSearch.categoriesLabel', { ns: 'homePage' });
    const localizedSortOrderLabel = i18n.t('bookSearch.sortOrderLabel', { ns: 'homePage' });
    const localizedSearchByLabel = i18n.t('bookSearch.searchByLabel', { ns: 'homePage' });

    render(
      <RouterProvider router={router} />,
    );

    expect(screen.getByLabelText<HTMLLabelElement>(localizedCategoriesLabel)).toBeInTheDocument();
    expect(screen.getByLabelText<HTMLLabelElement>(localizedSortOrderLabel)).toBeInTheDocument();
    expect(screen.getByLabelText<HTMLLabelElement>(localizedSearchByLabel)).toBeInTheDocument();
  });

  it('the component with the text in Russian is displayed correctly', () => {
    i18n.changeLanguage('ru');
    const localizedCategoriesLabel = i18n.t('bookSearch.categoriesLabel', { ns: 'homePage' });
    const localizedSortOrderLabel = i18n.t('bookSearch.sortOrderLabel', { ns: 'homePage' });
    const localizedSearchByLabel = i18n.t('bookSearch.searchByLabel', { ns: 'homePage' });

    render(
      <RouterProvider router={router} />,
    );

    expect(screen.getByLabelText<HTMLLabelElement>(localizedCategoriesLabel)).toBeInTheDocument();
    expect(screen.getByLabelText<HTMLLabelElement>(localizedSortOrderLabel)).toBeInTheDocument();
    expect(screen.getByLabelText<HTMLLabelElement>(localizedSearchByLabel)).toBeInTheDocument();
  });
});
