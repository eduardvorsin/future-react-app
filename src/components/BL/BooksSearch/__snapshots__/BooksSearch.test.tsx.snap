// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BooksSearch tests is a basic snapshot 1`] = `
<div
  class="books-search "
  data-testid="test-book-search"
>
  <form
    class="books-search__search-bar search-form search-form__search-bar"
  >
    <div
      class="search-bar"
      role="search"
    >
      <label
        class="search-bar__label sr-only"
        for="search-bar"
      >
        Search Bar
      </label>
      <input
        class="search-bar__input"
        id="search-bar"
        placeholder="Find a book..."
        value=""
      />
      <button
        class="button--icon-button search-bar__clear"
      >
        <svg />
        Clear
      </button>
      <button
        class="button--icon-button search-bar__find"
        type="submit"
      >
        <svg />
        To find
      </button>
    </div>
  </form>
  <div
    class="books-search__category-select books-search__select select"
  >
    <label
      class="select__label"
      for="categories"
    >
      Categories
    </label>
    <div
      class="select-wrapper"
    >
      <select
        class="select-wrapper__select"
        id="categories"
      >
        <option
          value="all"
        >
          All
        </option>
        <option
          value="art"
        >
          Art
        </option>
        <option
          value="biography"
        >
          Biography
        </option>
        <option
          value="computers"
        >
          Computers
        </option>
        <option
          value="history"
        >
          History
        </option>
        <option
          value="medical"
        >
          Medical
        </option>
        <option
          value="poetry"
        >
          Poetry
        </option>
      </select>
    </div>
  </div>
  <div
    class="books-search__sort-select books-search__select select"
  >
    <label
      class="select__label"
      for="sortingBy"
    >
      Sort by
    </label>
    <div
      class="select-wrapper"
    >
      <select
        class="select-wrapper__select"
        id="sortingBy"
      >
        <option
          value="relevance"
        >
          Relevance
        </option>
        <option
          value="newest"
        >
          Newest
        </option>
      </select>
    </div>
  </div>
  <div
    class="books-search__find-by-select books-search__select select"
  >
    <label
      class="select__label"
      for="findBy"
    >
      Search by
    </label>
    <div
      class="select-wrapper"
    >
      <select
        class="select-wrapper__select"
        id="findBy"
      >
        <option
          value="intitle"
        >
          By title
        </option>
        <option
          value="inauthor"
        >
          By author
        </option>
        <option
          value="inpublisher"
        >
          By publisher
        </option>
        <option
          value="isbn"
        >
          By ISBN
        </option>
      </select>
    </div>
  </div>
</div>
`;
